tests/foreman/cli/test_errata.py

```diff
@@ -623,6 +623,17 @@ def test_positive_list_affected_chosts_by_erratum_restrict_flag(
 
     :CaseAutomation: Automated
     """
+
+    @request.addfinalizer
+    def cleanup():
+        cv_filter_cleanup(
+            target_sat,
+            cv_filter['filter-id'],
+            module_cv,
+            module_sca_manifest_org,
+            module_lce,
+        )
```

Comment:
What would happen if the finalizer here was called before one of those arguments was created, cv_filter for example?

Reply:
yeah it is a confusing syntax compared to what we are normally used to in python. The function marked with `@request.addfinalizer` within the test function is *executed* **after** the test runs, so you can use variables even before you define them (!). But at the same time, the position of the block within the test determines when the finalizer is *added*. So I place these blocks at the start to be sure the finalizer is added at all times. The position in test matters for addition but not for execution

Reply:
And I think that's what Jacob asks about. What if the finalizer is added (here, at the start) and then something fails after it is added, but before e.g. `cv_filter` variable is defined. I suppose the finalizer would throw an exception?

Reply:
correct @lhellebr. at that point, you'd raise an exception because the name `cv_filter` is referenced before definition. Moving this to right after everything is defined is safer.

Reply:
@JacobCallahan @lhellebr I see, yeah I suppose it would raise an exception, but that's not different from "normal" finalizer behavior if defined in fixture. Consider the two alternatives when having finalizer in test body: 

a) tests fails in the middle, finalizer added before -> result: finalizer added and executed, exception raised if referencing stuff that didn't get created 
b) test fails in the middle, finalizer added after -> result: finalizer not added nor executed, not cleaning up anything that got created before failure, possibly causing interference with other tests

Situation (b) actually happens in the upgrade run, and it took considerable time and effort just to find out what is going on. Therefore I'd much rather have situation (a), which would be easier to debug